bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,395 @@
+        final String s = queryStringState(action, name, module);
+        try {
+            if (s == null) {
+                return UnDefType.UNDEF;
+            } else {
+                return new QuantityType<>(Double.valueOf(s), unit);
+            }
+        } catch (NumberFormatException e) {
+            logger.error("Unexpected number format: {}", s);
+        }
+        return UnDefType.UNDEF;
+    }
+
+    public void setState(final String action, final String name, final String moduleID, final String nick,
+            final String description, final OnOffType state) {
+        setState(action, name, moduleID, nick, description, OnOffType.ON.equals(state) ? "true" : "false");
+    }
+
+    public void setState(final String action, final String name, final String moduleID, final String nick,
+            final String description, final String state) {
+        try {
+            final SOAPMessage message = newMessage(action, moduleParameters(moduleID),
+                    soapFactory.createElement("NickName").addTextNode(nick),
+                    soapFactory.createElement("Description").addTextNode(description),
+                    soapFactory.createElement("OPStatus").addTextNode(state));
+            final Document document = post(action, message, authentication);
+            final String response = getElementByTagName(document, name);
+            if (!"ok".equals(response.toLowerCase())) {
+                logger.warn("Failed to set state.");
+            }
+        } catch (SOAPException e) {
+            logger.error("Unexpected internal error.");
+            updateStatus(ThingStatus.OFFLINE);
+        } catch (Exception e) {
+            // pass
+        }
+    }
+
+    private SOAPMessage newMessage(final String action, final SOAPElement... params) throws SOAPException {
+        if (soapMessageFactory == null) {
+            soapMessageFactory = MessageFactory.newInstance();
+        }
+        final SOAPMessage message = soapMessageFactory.createMessage();
+        message.getSOAPHeader().detachNode();
+        final SOAPElement actionElement = message.getSOAPBody().addChildElement(action, "",
+                "http://purenetworks.com/HNAP1/");
+        for (SOAPElement param : params) {
+            actionElement.addChildElement(param);
+        }
+        return message;
+    }
+
+    private SOAPElement newSOAPElement(final String name) throws SOAPException {
+        return newSOAPElement(name, null);
+    }
+
+    private SOAPElement newSOAPElement(final String name, final String text) throws SOAPException {
+        if (soapFactory == null) {
+            soapFactory = SOAPFactory.newInstance();
+        }
+        SOAPElement element = soapFactory.createElement(name);
+        if (text != null) {
+            element.addTextNode(text);
+        }
+        return element;
+    }
+
+    private boolean authenticate() {
```

Comment:
Sorry for sounding like a stuck record, but I don't see why this class needs to reimplement the low level HNAP communication. If this class extended `DLinkHNAPCommunication` then this method could be written as:

```
private boolean authenticate() {
        if (getHNAPStatus() == HNAPStatus.INVALID_PIN) {
            // If the configuration (i.e., PIN) is invalid, trying to long in does not make sense.
            return false;
        } else if (getHNAPStatus() == HNAPStatus.LOGGED_IN) {
            // We are already authenticated, nothing to do.
            return true;
        }

        login(5);

        if (getHNAPStatus() == HNAPStatus.LOGGED_IN) {
            updateStatus(ThingStatus.ONLINE);
        } else if (getHNAPStatus() == HNAPStatus.INVALID_PIN) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Failed to authenticate.");
        } else {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Failed to authenticate.");
        }
    }
```

The post method could be written as:
```
private Document post(final String action, final SOAPMessage message)
            throws CommunicationException {

        final MimeHeaders headers = message.getMimeHeaders();
        final String actionUrl = "\"http://purenetworks.com/HNAP1/" + action + "\"";
        headers.addHeader(SOAPACTION, actionUrl);

        setAuthenticationHeaders(message);
        
        try {
            return sendReceive(message, 5);
        } catch (InterruptedException | TimeoutException | ExecutionException | SAXException | IOException e) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
            throw new CommunicationException("Post request failed: " + e.getMessage());
        }
    }
```
All the exceptions thrown from sendReceive look horrible, so you could add a new method to `DLinkHNAPCommunication` which catches all those exceptions and sets the status to COMMUNICATION_ERROR.

It might also make sense to add `getElementByTagName` to `DLinkHNAPCommunication` which sets the status to COMMUNICATION_ERROR if the requested tag isn't present.

Some minor changes to  `DLinkHNAPCommunication` would be required to use the common httpclient rather than creating its own.

Reply:
Again: There is no need to reimplement the `DLinkHNAPCommunication`, but it needs improvement, imho.
I wrote the new implementation to learn, not because it was required. Now we have it for free.
Eventually, there should be only one implementation, I don't care if you (or I) improve mine or yours.

Reply:
I'm trying to suggest some changes which make use of existing code and removes the need to maintain two versions of code which are doing the same thing.

Switching over the motion detector to use your `DLinkHNAP` class looks like it will involve significant changes and means we have to thoroughly test two devices rather than just the new device.

The changes I've suggested looks like it removes duplicate code from `DLinkHNAP` and doesn't change the interface to your `DLinkSmartPlugHandler` class. The existing HNAP implementation has had many years of production use, can you elaborate on what improvements you feel your implementation of `authenticate` and `post` have made compared to what I have suggested which calls the existing code?

If we are going to live with two implementations of the HNAP communication then perhaps we just package them up so that they belong to their respective devices, i.e. move `DLinkHNAPCommunication` into motionSensor and move `DLinkHNAP` into smartPlug.